File: tup/tup-csharp/tup/UniAttribute.cs
/**
 * Tencent is pleased to support the open source community by making Tars available.
 *
 * Copyright (C) 2016THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except 
 * in compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed 
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR 
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the 
 * specific language governing permissions and limitations under the License.
 */


using Tup.Tars;

using System;
using System.Collections.Generic;
using System.Text;
using System.Collections;

namespace Tup
{
    public class UniAttribute : TarsStruct
    {
        /**
         * PACKET_TYPE_TUP类型
         */
        protected Dictionary<string, Dictionary<string, byte[]>> _data = null;
        /**
         * 精简版tup，PACKET_TYPE_TUP3类型
         */
        protected Dictionary<string, byte[]> _new_data = null;

        /**
         * 存储get后的数据 避免多次解析
         */
        private Dictionary<string, object> cachedData = new Dictionary<string, object>(128);

        protected short _iVer = Const.PACKET_TYPE_TUP;
        public short Version
        {
            get
            {
                return _iVer;
            }
            set
            {
                _iVer = value;
            }
        }

        private string _encodeName = "UTF-8";
        public string EncodeName
        {
            get
            {
                return _encodeName;
            }
            set
            {
                _encodeName = value;
            }
        }

        TarsInputStream _is = new TarsInputStream();

        public UniAttribute()
        {
            _data = new Dictionary<string, Dictionary<string, byte[]>>();
            _new_data = new Dictionary<string, byte[]>();
        }

        /**
         * 清除缓存的解析过的数据
         */
        public void ClearCacheData()
        {
            cachedData.Clear();
        }

        public bool IsEmpty()
        {
            if (_iVer == Const.PACKET_TYPE_TUP3)
            {
                return _new_data.Count == 0;
            }
            else
            {
                return _data.Count == 0;
            }
        }

        public int Size
        {
            get
            {
                if (_iVer == Const.PACKET_TYPE_TUP3)
                {
                    return _new_data.Count;
                }
                else
                {
                    return _data.Count;
                }
            }
        }

        public bool ContainsKey(string key)
        {
            if (_iVer == Const.PACKET_TYPE_TUP3)
            {
                return _new_data.ContainsKey(key);
            }
            else
            {
                return _data.ContainsKey(key);
            }
        }

        /**
         * 放入一个元素
         * @param <T>
         * @param name
         * @param t
         */
        public void Put<T>(string name, T t)
        {
            if (name == null)
            {
                throw new ArgumentException("put key can not is null");
            }
            if (t == null)
            {
                throw new ArgumentException("put value can not is null");
            }

            TarsOutputStream _out = new TarsOutputStream();
            _out.setServerEncoding(_encodeName);
            _out.Write(t, 0);
			byte[] sBuffer = TarsUtil.getTarsBufArray(_out.getMemoryStream());

            if (_iVer == Const.PACKET_TYPE_TUP3)
            {
                cachedData.Remove(name);

                if (_new_data.ContainsKey(name))
                {
                    _new_data[name] = sBuffer;
                }
                else
                {
                    _new_data.Add(name, sBuffer);
                }
            }
            else
            {
                List<string> listType = new List<string>();
                CheckObjectType(listType, t);
                string className = BasicClassTypeUtil.TransTypeList(listType);

                Dictionary<string, byte[]> pair = new Dictionary<string, byte[]>(1);
                pair.Add(className, sBuffer);
                cachedData.Remove(name);

                if (_data.ContainsKey(name))
                {
                    _data[name] = pair;
                }
                else
                {
                    _data.Add(name, pair);
                }
            }
        }

        private Object decodeData(byte[] data, Object proxy)
        {
            _is.wrap(data);
            _is.setServerEncoding(_encodeName);
            Object o = _is.read(proxy, 0, true);
            return o;
        }

        /**
         * 获取tup精简版本编码的数据,兼容旧版本tup
         * @param <T>
         * @param name
         * @param proxy
         * @return
         * @throws ObjectCreateException
         */
        public T getByClass<T>(string name, T proxy)
        {
            object obj = null;
            if(_iVer == Const.PACKET_TYPE_TUP3)
            {

                if (!_new_data.ContainsKey(name))
                {
                    return (T)obj;
                }
                else if (cachedData.ContainsKey(name))
                {
                    if (!cachedData.TryGetValue(name, out obj))
                    {
                        obj = null;
                    }
                    return (T)obj;
                }
                else
                {
                    try
                    {
                        byte[] data = new byte[0];
                        _new_data.TryGetValue(name, out data);

                        Object o = decodeData(data, proxy);
                        if (null != o)
                        {
                            SaveDataCache(name, o);
                        }
                        return (T)o;
                    }
                    catch (Exception ex)
                    {
                        throw new ObjectCreateException(ex);
                    }
                }
            }
            else //兼容tup2
            {
                return Get<T>(name);
            }
        }


        /**
         * 获取一个元素,只能用于tup版本2，如果待获取的数据为tup3，则抛异常
         * @param <T>
         * @param name
         * @return
         * @throws ObjectCreateException
         */
        public T Get<T>(string name)
        {
            if (_iVer == Const.PACKET_TYPE_TUP3)
            {
                throw new Exception("data is encoded by new version, please use getTarsStruct(String name,T proxy)");
            }

            object obj = null;

            if (!_data.ContainsKey(name))
            {
                return (T)obj;
            }
            else if (cachedData.ContainsKey(name))
            {
                if (!cachedData.TryGetValue(name, out obj))
                {
                    obj = null;
                }
                return (T)obj;
            }
            else
            {
                Dictionary<string, byte[]> pair;
                _data.TryGetValue(name, out pair);

                string strBasicType = "";
                string className = null;
                byte[] data = new byte[0];

                // 找到和T类型对应的数据data
                foreach (KeyValuePair<string, byte[]> e in pair)
                {
                    className = e.Key;
                    data = e.Value;

                    if (className == null || className == string.Empty)
                    {
                        continue;
                    }

                    // 比较基本类型
                    strBasicType = BasicClassTypeUtil.CS2UniType(typeof(T).ToString());
                    if (className.Length > 0 && className == strBasicType)
                    {
                        break;
                    }
                    if (strBasicType == "map" && className.Length >= 3 && className.Substring(0, 3).ToLower() == "map")
                    {
                        break;
                    }
                    if (typeof(T).IsArray && className.Length > 3 && className.Substring(0, 4).ToLower() == "list")
                    {
                        break;
                    }
                    if (strBasicType == "list" && className.Length > 3 && className.Substring(0, 4).ToLower() == "list")
                    {
                        break;
                    }
                }

                try
                {
                    object objtmp = GetCacheProxy<T>(className);
                    if (objtmp == null)
                    {
                        return (T)objtmp;
                    }
       
                    obj = decodeData(data, objtmp);
                    if (obj != null)
                    {
                        SaveDataCache(name, obj);
                    }
                    return (T)obj;
                }
                catch (Exception ex)
                {
                    QTrace.Trace(this + " Get Exception: " + ex.Message);
                    throw new ObjectCreateException(ex);
                }
            }
        }

        /**
          * 获取一个元素,tup新旧版本都兼容
          * @param Name
          * @param DefaultObj
          * @return
          * @throws ObjectCreateException
          */
        public T Get<T>(string Name, T DefaultObj)
        {
            try
            {
                object result = null; ;
             
                if (_iVer == Const.PACKET_TYPE_TUP3)
                {
                    result = getByClass<T>(Name, DefaultObj);
                    
                }
                else //tup2
                {
                    result = Get<T>(Name);
                }
                
                if (result == null)
                {
                    return DefaultObj;
                }
                return (T)result;
            }
            catch
            {
                return DefaultObj;
            }
        }

        private Object GetCacheProxy<T>(string className)
        {
            return BasicClassTypeUtil.CreateObject<T>();
        }

        private void SaveDataCache(string name, Object o)
        {
            cachedData.Add(name, o);
        }

        /**
         * 检测传入的元素类型
         * 
         * @param listTpye
         * @param o
         */
        private void CheckObjectType(List<string> listTpye, Object o)
        {
            if (o == null)
            {
                throw new Exception("object is null");
            }

            if (o.GetType().IsArray)
            {
                Type elementType = o.GetType().GetElementType();
                listTpye.Add("list");
                CheckObjectType(listTpye, BasicClassTypeUtil.CreateObject(elementType));
            }
            else if (o is IList)
            {
                listTpye.Add("list");

                IList list = (IList)o;
                if (list.Count > 0)
                {
                    CheckObjectType(listTpye, list[0]);
                }
                else
                {
                    listTpye.Add("?");
                }
            }
            else if (o is IDictionary)
            {
                listTpye.Add("map");
                IDictionary map = (IDictionary)o;
                if (map.Count > 0)
                {
                    foreach (object key in map.Keys)
                    {
                        listTpye.Add(BasicClassTypeUtil.CS2UniType(key.GetType().ToString()));
                        CheckObjectType(listTpye, map[key]);
                        break;
                    }
                }
                else
                {
                    listTpye.Add("?");
                    listTpye.Add("?");
                    //throw new ArgumentException("map  can not is empty");
                }
            }
            else
            {
                listTpye.Add(BasicClassTypeUtil.CS2UniType(o.GetType().ToString()));
            }
        }

        public byte[] Encode()
        {
            TarsOutputStream _os = new TarsOutputStream(0);
            _os.setServerEncoding(_encodeName);
            if (_iVer == Const.PACKET_TYPE_TUP3)
            {
                _os.Write(_new_data, 0);
            }
            else
            {
                _os.Write(_data, 0);
            }
            return TarsUtil.getTarsBufArray(_os.getMemoryStream());
        }

        public void Decode(byte[] buffer, int Index = 0)
        {
            try
            {
                //try tup2
                _is.wrap(buffer, Index);
                _is.setServerEncoding(_encodeName);
                _iVer = Const.PACKET_TYPE_TUP;
                _data = (Dictionary<string, Dictionary<string, byte[]>>)_is.readMap<Dictionary<string, Dictionary<string, byte[]>>>(_data,0, false);
                return;
            }catch
            {
                //try tup3
                _iVer = Const.PACKET_TYPE_TUP3;
                _is.wrap(buffer, Index);
                _is.setServerEncoding(_encodeName);
                _new_data = (Dictionary<string, byte[]>)_is.readMap<Dictionary<string, byte[]>>(_new_data,0, false);     
            }
        }

        public override void WriteTo(TarsOutputStream _os)
        {
            if (_iVer == Const.PACKET_TYPE_TUP3)
            {
                _os.Write(_new_data, 0);
            }
            else
            {
                _os.Write(_data, 0);
            }
        }

        public override void ReadFrom(TarsInputStream _is)
        {
            if (_iVer == Const.PACKET_TYPE_TUP3)
            {
                _new_data = (Dictionary<string, byte[]>)_is.readMap<Dictionary<string, byte[]>>(_new_data, 0, false); 
            }
            else
            {
                _data = (Dictionary<string, Dictionary<string, byte[]>>)_is.readMap<Dictionary<string, Dictionary<string, byte[]>>>(_data, 0, false);
            }
        }
    }
}

File: Tars/java/core/src/main/java/com/qq/tars/rpc/protocol/tup/UniAttribute.java
/**
 * Tencent is pleased to support the open source community by making Tars available.
 *
 * Copyright (C) 2016 THL A29 Limited, a Tencent company. All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software distributed
 * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package com.qq.tars.rpc.protocol.tup;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map.Entry;
import java.util.Set;

import com.qq.tars.protocol.tars.TarsInputStream;
import com.qq.tars.protocol.tars.TarsOutputStream;
import com.qq.tars.protocol.util.TarsUtil;

public class UniAttribute extends OldUniAttribute {

    protected HashMap<String, byte[]> _newData = new HashMap<String, byte[]>();

    private HashMap<String, Object> cachedData = new HashMap<String, Object>();

    TarsInputStream _is = new TarsInputStream();

    public void useVersion3() {
        _newData = new HashMap<String, byte[]>();
    }

    public void clearCacheData() {
        cachedData.clear();
    }

    public Set<String> getKeySet() {
        if (null != _newData) {
            return Collections.unmodifiableSet(_newData.keySet());
        }
        return Collections.unmodifiableSet(_data.keySet());
    }

    public boolean isEmpty() {
        if (null != _newData) {
            return _newData.isEmpty();
        }
        return _data.isEmpty();
    }

    public int size() {
        if (null != _newData) {
            return _newData.size();
        }
        return _data.size();
    }

    public boolean containsKey(String key) {
        if (null != _newData) {
            return _newData.containsKey(key);
        }
        return _data.containsKey(key);
    }

    public <T> void put(String name, T t) {
        if (null != _newData) {
            if (name == null) {
                throw new IllegalArgumentException("put key can not is null");
            }
            if (t == null) {
                throw new IllegalArgumentException("put value can not is null");
            }
            if (t instanceof Set) {
                throw new IllegalArgumentException("can not support Set");
            }
            TarsOutputStream _out = new TarsOutputStream();
            _out.setServerEncoding(encodeName);
            _out.write(t, 0);
            byte[] _sBuffer = TarsUtil.getBufArray(_out.getByteBuffer());
            _newData.put(name, _sBuffer);
        } else {
            super.put(name, t);
        }
    }

    @SuppressWarnings("unchecked")
    public <T> T getStruct(String name, T proxy) throws ObjectCreateException {
        if (!_newData.containsKey(name)) {
            return null;
        } else if (cachedData.containsKey(name)) {
            return (T) cachedData.get(name);
        } else {
            byte[] data = _newData.get(name);
            try {
                Object o = decodeData(data, proxy);
                if (null != o) {
                    saveDataCache(name, o);
                }
                return (T) o;
            } catch (Exception ex) {
                throw new ObjectCreateException(ex);
            }
        }
    }

    public <T> T getByClass(String name, T proxy) throws ObjectCreateException {
        if (null != _newData) {
            if (!_newData.containsKey(name)) {
                return null;
            } else if (cachedData.containsKey(name)) {
                return (T) cachedData.get(name);
            } else {
                byte[] data = _newData.get(name);
                try {
                    Object o = decodeData(data, proxy);
                    if (null != o) {
                        saveDataCache(name, o);
                    }
                    return (T) o;
                } catch (Exception ex) {
                    throw new ObjectCreateException(ex);
                }
            }
        } else {
            if (!_data.containsKey(name)) {
                return null;
            } else if (cachedData.containsKey(name)) {
                return (T) cachedData.get(name);
            } else {
                HashMap<String, byte[]> pair = _data.get(name);
                String className = null;
                byte[] data = new byte[0];
                for (Entry<String, byte[]> e : pair.entrySet()) {
                    className = e.getKey();
                    data = e.getValue();
                    break;
                }
                try {
                    _is.warp(data);
                    _is.setServerEncoding(encodeName);
                    Object o = _is.read(proxy, 0, true);
                    saveDataCache(name, o);
                    return (T) o;
                } catch (Exception ex) {
                    throw new ObjectCreateException(ex);
                }
            }
        }

    }

    public <T> T getByClass(String name, T proxy, T defaultValue) throws ObjectCreateException {
        if (null != _newData) {
            if (!_newData.containsKey(name)) {
                return defaultValue;
            } else if (cachedData.containsKey(name)) {
                return (T) cachedData.get(name);
            } else {
                byte[] data = _newData.get(name);
                try {
                    Object o = decodeData(data, proxy);
                    if (null != o) {
                        saveDataCache(name, o);
                    }
                    return (T) o;
                } catch (Exception ex) {
                    throw new ObjectCreateException(ex);
                }
            }
        } else {
            if (!_data.containsKey(name)) {
                return defaultValue;
            } else if (cachedData.containsKey(name)) {
                return (T) cachedData.get(name);
            } else {
                HashMap<String, byte[]> pair = _data.get(name);
                String className = null;
                byte[] data = new byte[0];
                for (Entry<String, byte[]> e : pair.entrySet()) {
                    className = e.getKey();
                    data = e.getValue();
                    break;
                }
                try {
                    _is.warp(data);
                    _is.setServerEncoding(encodeName);
                    Object o = _is.read(proxy, 0, true);
                    saveDataCache(name, o);
                    return (T) o;
                } catch (Exception ex) {
                    throw new ObjectCreateException(ex);
                }
            }
        }

    }

    @SuppressWarnings("unchecked")
    public <T> T get(String name, T proxy, Object defaultValue) {
        if (!_newData.containsKey(name)) {
            return (T) defaultValue;
        }
        return getByClass(name, proxy);
    }

    private Object decodeData(byte[] data, Object proxy) {
        _is.warp(data);
        _is.setServerEncoding(encodeName);
        Object o = _is.read(proxy, 0, true);
        return o;
    }

    private void saveDataCache(String name, Object o) {
        cachedData.put(name, o);
    }

    public <T> T getStruct(String name) throws ObjectCreateException {
        if (null != _newData) {
            throw new RuntimeException("data is encoded by new version, please use getStruct(String name,T proxy)");
        } else {
            return (T) super.getStruct(name);
        }
    }

    public <T> T get(String name) throws ObjectCreateException {
        if (null != _newData) {
            throw new RuntimeException("data is encoded by new version, please use getByClass(String name, T proxy)");
        } else {
            return (T) super.get(name);
        }
    }

    public <T> T get(String name, Object defaultValue) {
        if (null != _newData) {
            throw new RuntimeException("data is encoded by new version, please use get(String name, T proxy, Object defaultValue)");
        } else {
            return (T) super.get(name, defaultValue);
        }
    }

    public <T> T remove(String name) throws ObjectCreateException {
        if (null != _newData) {
            if (!_newData.containsKey(name)) {
                return null;
            } else {
                _newData.remove(name);
                return null;
            }
        } else {
            return (T) super.remove(name);
        }

    }

    @SuppressWarnings("unchecked")
    public <T> T remove(String name, T proxy) throws ObjectCreateException {
        if (!_newData.containsKey(name)) {
            return null;
        } else {
            if (null != proxy) {
                byte[] data = _newData.remove(name);
                Object o = decodeData(data, proxy);
                return (T) o;
            } else {
                _newData.remove(name);
                return null;
            }
        }
    }

    public byte[] encode() {
        if (null != _newData) {
            TarsOutputStream _os = new TarsOutputStream(0);
            _os.setServerEncoding(encodeName);
            _os.write(_newData, 0);
            return TarsUtil.getBufArray(_os.getByteBuffer());
        }
        return super.encode();
    }

    public void decodeVersion3(byte[] buffer) {
        _is.warp(buffer);
        _is.setServerEncoding(encodeName);
        HashMap<String, byte[]> _tempdata = new HashMap<String, byte[]>(1);
        _tempdata.put("", new byte[0]);
        _newData = (HashMap<String, byte[]>) _is.readMap(_tempdata, 0, false);
    }

    public void decodeVersion2(byte[] buffer) {
        super.decode(buffer);
    }

    public void decode(byte[] buffer) {
        try {
            super.decode(buffer);
            return;
        } catch (Exception e) {
            _is.warp(buffer);
            _is.setServerEncoding(encodeName);
            HashMap<String, byte[]> _tempdata = new HashMap<String, byte[]>(1);
            _tempdata.put("", new byte[0]);
            _newData = (HashMap<String, byte[]>) _is.readMap(_tempdata, 0, false);
        }
    }

}
